common/posix.edl

```diff
@@ -0,0 +1,395 @@
+            size_t count,
+            [out, count=1] int* err);
+
+        oe_off_t oe_posix_lseek_ocall(
+            oe_host_fd_t fd,
+            oe_off_t offset,
+            int whence,
+            [out, count=1] int* err);
+
+        int oe_posix_close_ocall(
+            oe_host_fd_t fd,
+            [out, count=1] int* err);
+
+        oe_host_fd_t oe_posix_dup_ocall(
+            oe_host_fd_t oldfd,
+            [out, count=1] int* err);
+
+        uint64_t oe_posix_opendir_ocall(
+            [in, string] const char* name,
+            [out, count=1] int* err);
+
+        int oe_posix_readdir_ocall(
+            uint64_t dirp,
+            [out, count=1] uint64_t* d_ino,
+            [out, count=1] int64_t* d_off,
+            [out, count=1] uint16_t* d_reclen,
+            [out, count=1] uint8_t* d_type,
+            [out, count=d_namelen] char* d_name,
+            size_t d_namelen,
+            [out, count=1] int* err);
+
+        void oe_posix_rewinddir_ocall(
+            uint64_t dirp);
+
+        int oe_posix_closedir_ocall(
+            uint64_t dirp,
+            [out, count=1] int* err);
+
+        int oe_posix_stat_ocall(
+            [in, string] const char* pathname,
+            [out, count=1] struct oe_stat* buf,
+            [out, count=1] int* err);
+
+        int oe_posix_access_ocall(
+            [in, string] const char* pathname,
+            int mode,
+            [out, count=1] int* err);
+
+        int oe_posix_link_ocall(
+            [in, string] const char* oldpath,
+            [in, string] const char* newpath,
+            [out, count=1] int* err);
+
+        int oe_posix_unlink_ocall(
+            [in, string] const char* pathname,
+            [out, count=1] int* err);
+
+        int oe_posix_rename_ocall(
+            [in, string] const char* oldpath,
+            [in, string] const char* newpath,
+            [out, count=1] int* err);
+
+        int oe_posix_truncate_ocall(
+            [in, string] const char* path,
+            oe_off_t length,
+            [out, count=1] int* err);
+
+        int oe_posix_mkdir_ocall(
+            [in, string] const char* pathname,
+            oe_mode_t mode,
+            [out, count=1] int* err);
+
+        int oe_posix_rmdir_ocall(
+            [in, string] const char* pathname,
+            [out, count=1] int* err);
+
+       oe_host_fd_t oe_posix_socket_ocall(
+            int domain,
+            int type,
+            int protocol,
+            [out, count=1] int* err);
+
+        int oe_posix_socketpair_ocall(
+            int domain,
+            int type,
+            int protocol,
+            [in] oe_host_fd_t sv[2],
+            [out, count=1] int* err);
+
+        int oe_posix_connect_ocall(
+            oe_host_fd_t sockfd,
+            [in, size=addrlen] const struct oe_sockaddr* addr,
+            oe_socklen_t addrlen,
+            [out, count=1] int* err);
+
+        oe_host_fd_t oe_posix_accept_ocall(
```

Comment:
@andschwa and @anakrish: This function defines **addrlen_in** and **addrlen_out**. Is there a way to get by with only one parameter. The following does not seem to work since addlen is a pointer and not an integer.

```
oe_host_fd_t oe_posix_accept_ocall(
     oe_host_fd_t sockfd,
     [in, out, size=addrlen] struct oe_sockaddr* addr,
     [in, out, count=1] oe_socklen_t* addrlen,
     [out, count=1] int* err);
```

Reply:
@mikbras Your alternative using in/out parameter should work. Most likely an edger8r bug if it does not.
I don't see how your example is different to:
https://github.com/Microsoft/openenclave/blob/ada28f14b4b8f5fdd1450c6e41e12db2cb09248e/tests/oeedger8r/edl/foreign.edl#L15-L19

Reply:
It is a bug. I created issue #1748 to track.

What is different is that the size argument is a pointer to a size, not an actual size.

Reply:
Ah ok. I see it now. It is currently not supported by edger8r. This requirement of an in-out size parameter has cropped up a few times in the past. Would be worth considering if EDL ought to be enhanced to support this scenario.

Reply:
It works fine in EDL. It is the emitter that generates incorrect code.

Reply:
Yes.It is syntactically valid in EDL.
Just like _int float a;_ is syntactically valid but semantically invalid C code. 

EDL currently does not claim to support our example; however it does not emit an _unsupported construct_ error message either. @andschwa we can start off by erroring out and following up with intel if we want to enhance the semantics.

Reply:
@mikbras the upstream EDL parser we use from Intel does not really do much in the way of validation, so there is quite a lot of "valid EDL" code you can write that will parse fine and result in silent bugs. We have some additional warnings around count/size parameters implemented in our edger8r that I'd like to ship back upstream to Intel, and checking if pointers are being used erroneously would be a great extra warning to add until if/when this is better supported.

Reply:
Ah I see. Thanks Andy for the update.